### PR TITLE
Fix undefined behavior after changing HMICapabilities smart object field

### LINKED ISSUE #3109: SDL Crash - App registration response
### Bug Report
SDL crashed during app registration

##### Reproduction Steps
1. Make sure SDL is built in `EXTERNAL_PROPRIETARY` mode
2. Run SDL, HMI, connect Mobile
3. Register new app

##### Expected Behavior
App registered successfully

##### Observed Behavior
SDL crashed during sending of RAI response

##### OS & Version Information
* OS/Version: Ubuntu 16.04
* SDL Core Version: https://github.com/smartdevicelink/sdl_core/tree/develop (9ac782b4)
* Testing Against: ATF 

##### Attachments
  - script: [144_ATF_PTU_validation_failure](https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/develop/test_scripts/Policies/Policy_Table_Update/144_ATF_PTU_validation_failure.lua)
  - reports and logs: [atf_sdl_logs.tar.gz](https://github.com/smartdevicelink/sdl_core/files/3794410/atf_sdl_logs.tar.gz)
  - corefile: [corefiles.tar.gz](https://github.com/smartdevicelink/sdl_core/files/3793570/corefiles.tar.gz)
  - gdb: [gdb (8).txt](https://github.com/smartdevicelink/sdl_core/files/3793571/gdb.8.txt)

### PR DESCRIPTION
**Pull request by AKalinich-Luxoft**
Fixes #3109 

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
Covered by ATF tests

### Summary
After `cleanup_data()` call, delete is used to deallocate memory by pointer, however pointer still hold a value of address of deallocated memory. By that reason SO still have an ability to access that data what might cause UB including core crash in the random places.

This function has been updated to set pointers to NULL as well as smart object type to prevent accident access to deallocated memory.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/smart_objects/src/smart_object.cc [hunk 1]
```diff
@@ -734,16 +734,32 @@ void SmartObject::duplicate(const SmartObject& OtherObject) {
...
+        delete m_data.array_value;
+        m_data.array_value = nullptr;
+        m_type = SmartType_Null;
+      }
       break;
     case SmartType_Binary:
-      delete m_data.binary_value;
+      if (m_data.binary_value) {
+        delete m_data.binary_value;
+        m_data.array_value = nullptr;
```
Comment (JackLivio) [inline src/components/smart_objects/src/smart_object.cc]: Copy paste error

**Comment by iCollin** [comment]
As far as I can tell, `cleanup_data` only deals with `SmartObject::m_data` (and in this branch `SmartObject::m_type`), while the crash described in the issue happens while deleting `SmartObject::m_schema`.

`~SmartObject` will go on to destroy `m_schema` which contains the (in the case of the issue, invalid) `ISchemaItemPtr` after `cleanup_data` has returned. I don't understand how an invalid `ISchemaItemPtr` would be caused by code within `cleanup_data`.

**Comment by JackLivio** [comment]
Does this fix isue #3108 instead of #3109 ?

File: src/components/smart_objects/src/smart_object.cc (hunk 1 above)
Comment (AKalinich-Luxoft) [reply src/components/smart_objects/src/smart_object.cc]: @JackLivio fixed in https://github.com/smartdevicelink/sdl_core/pull/3116/commits/0ecdff2aa896168e0e57f04bb31d48cd955ac67d

**Comment by AKalinich-Luxoft** [comment]
@JackLivio no, 3108 has a quite different root cause and will have a separate fix

**Comment by AKalinich-Luxoft** [comment]
@iCollin PR is ready for Livio team review

**Pull request merged**